crates/server/src/store/blocklist/authority.rs

```diff
@@ -0,0 +1,689 @@
+};
+
+// TODO:
+//  * Add (optional) support for logging the client IP address.  This will require some Authority
+//    trait changes to accomplish.
+//  * Add query-type specific results for non-address queries.
+//  * Add support for per-blocklist sinkhole IPs, block messages, actions.
+//  * Add support for an exclusion list: allow the user to configure a list of patterns that
+//    will never be insert into the in-memory blocklist (such as their own domain.)
+//  * Add support for regex matching
+
+/// A conditional authority that will resolve queries against one or more block lists and return a
+/// forged response.  The typical use case will be to use this in a chained configuration before a
+/// forwarding or recursive resolver in order to pre-emptively block queries for hosts that are on
+/// a block list. Refer to tests/test-data/test_configs/chained_blocklist.toml for an example
+/// of this configuration.
+///
+/// The blocklist authority also supports the consult interface, which allows an authority to review
+/// a query/response that has been processed by another authority, and, optionally, overwrite that
+/// response before returning it to the requestor.  There is an example of this configuration in
+/// tests/test-data/test_configs/example_consulting_blocklist.toml.  The main intended use of this
+/// feature is to allow log-only configurations, to allow administrators to see if blocklist domains
+/// are being queried.  While this can be configured to overwrite responses, it is not recommended
+/// to do so - it is both more efficient, and more secure, to allow the blocklist to drop queries
+/// pre-emptively, as in the first example.
+pub struct BlocklistAuthority {
+    origin: LowerName,
+    blocklist: HashMap<LowerName, bool>,
+    wildcard_match: bool,
+    min_wildcard_depth: u8,
+    sinkhole_ipv4: Ipv4Addr,
+    sinkhole_ipv6: Ipv6Addr,
+    ttl: u32,
+    block_message: Option<String>,
+    consult_action: BlocklistConsultAction,
+}
+
+impl BlocklistAuthority {
+    /// Read the Authority for the origin from the specified configuration
+    pub async fn try_from_config(
+        origin: Name,
+        _zone_type: ZoneType,
+        config: &BlocklistConfig,
+        base_dir: Option<&Path>,
+    ) -> Result<Self, String> {
+        info!("loading blocklist config: {origin}");
+
+        let sinkhole_ipv4 = Ipv4Addr::from_str(&config.sinkhole_ipv4[..]).unwrap_or_else(|_| {
+            info!(
+                "invalid IPv4 sinkhole address '{}' provided; using default 0.0.0.0",
+                &config.sinkhole_ipv4[..]
+            );
+            Ipv4Addr::new(0, 0, 0, 0)
+        });
+        let sinkhole_ipv6 = Ipv6Addr::from_str(&config.sinkhole_ipv6[..]).unwrap_or_else(|_| {
+            info!(
+                "invalid IPv6 sinkhole address '{}' provided; using default ::0",
+                &config.sinkhole_ipv6[..]
+            );
+            Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 0)
+        });
```

Comment:
Sorry, but I find this setup of logging and then falling back to some default pretty bad. If someone is providing explicit configuration and it turns out to be invalid, we should yield an error for that so that they might correct it.

Reply:
how is this in any way different than making list io errors "fallible"?  The lists are the more important configuration element.

I think making either survivable is a bad design.

Reply:
I agree but I think the mechanism should be yielding a `Result::Err`, not just panicking, allowing the caller to deal with the error.

Reply:
I went with returning an error for now. IMO, if we're going to be this picky about server initialization code for the authorities, then all of that code should be moved from the server crate to the server binary source.  Propagating errors from authority methods, through the dyn AuthorityObject interface, and back to the server only to then abort for startup errors where we know the right decision is to abort is not a good design. We're losing important context (useful stack traces, for one thing,) for no real benefit.

Reply:
I'm sorry if my reviews are frustrating to deal with. I'm just trying to make the library code here easy to work with from the perspective of a library consumer -- and I do think it makes sense to provide a (good) library API for this code as well. Given that I'm reviewing in the browser, I may also miss some of this context, so thanks for pointing out this perspective to me.

I also think for many users of the binary, being confronted with a panic error message is not actually a good experience, and I think we could do better. I agree that it is important to capture enough context that makes it easy to understand why the error happens, and am happy to discuss how we could facilitate that (maybe using `anyhow::Context` could help for this stuff).

I'm not entirely sure how the `AuthorityObject` interface is in play during initialization, though? Looking at the changes here, it doesn't seem to be materially involved?

Reply:
Sure -- I think my recent PRs have shown that I'm in favor of moving code around between crates where it makes sense. The question is where exactly the right boundary might be for the blocklist authority.

Reply:
I'm going to mark this as resolved.  I think the blocklist configuration is consistent with what the other authorities are doing now, and whatever changes we make in the future, if any, should be consistent across all of the authorities.